server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -274,96 +272,14 @@ long buildTookInMillis() {
 
     @Override
     protected void doExecute(Task task, SearchRequest searchRequest, ActionListener<SearchResponse> listener) {
-        executeRequest((SearchTask) task, searchRequest, this::searchAsyncAction, listener);
+        executeRequest((SearchTask) task, searchRequest, listener, AsyncSearchActionProvider::new);
     }
 
-    public interface SinglePhaseSearchAction {
-        void executeOnShardTarget(
-            SearchTask searchTask,
-            SearchShardIterator shardIt,
-            Transport.Connection connection,
-            ActionListener<SearchPhaseResult> listener
-        );
-    }
-
-    public void executeRequest(
-        SearchTask task,
-        SearchRequest searchRequest,
-        String actionName,
-        boolean includeSearchContext,
-        SinglePhaseSearchAction phaseSearchAction,
-        ActionListener<SearchResponse> listener
```

Comment:
I need to remove the `ActionListener<SearchResponse> listener`parameter so we can use this action with the `can_match` phase only, which doesn't return a SearchResponse. This change is purely a matter of wiring things up.

Reply:
thanks a lot for extracting this out. It makes reviewing much simpler.